File: app/platform/runtime/redis_runtime.py
"""Redis-backed runtime primitives.

These helpers use Redis as an optional coordination layer. They are deliberately
small: callers keep their local hot path and opt into Redis for cross-process
state such as locks, counters, task snapshots, and short-lived cache entries.
"""

from __future__ import annotations

import os
import socket
import uuid
from typing import Any


def _decode(value: Any) -> str:
    return value.decode() if isinstance(value, bytes) else str(value)


class RedisRuntimeLease:
    """A named Redis lock lease."""

    def __init__(self, redis: Any, key: str, owner: str, ttl_ms: int) -> None:
        self._redis = redis
        self.key = key
        self.owner = owner
        self.ttl_ms = max(1, int(ttl_ms))
        self._released = False

    async def renew(self) -> bool:
        if self._released:
            return False
        current = await self._redis.get(self.key)
        if current is None or _decode(current) != self.owner:
            return False
        await self._redis.expire(self.key, max(1, self.ttl_ms // 1000))
        return True

    async def release(self) -> bool:
        if self._released:
            return False
        current = await self._redis.get(self.key)
        if current is None or _decode(current) != self.owner:
            self._released = True
            return False
        await self._redis.delete(self.key)
        self._released = True
        return True


class RedisRuntimeStore:
    """Small Redis runtime facade for locks and future shared state."""

    def __init__(self, redis: Any, *, key_prefix: str = "runtime:") -> None:
        self.redis = redis
        self.key_prefix = key_prefix

    def key(self, *parts: str) -> str:
        safe = ":".join(str(part).strip(":") for part in parts if str(part))
        return f"{self.key_prefix}{safe}"

    async def acquire_lock(
        self,
        name: str,
        *,
        owner: str | None = None,
        ttl_ms: int = 300_000,
    ) -> RedisRuntimeLease | None:
        owner = owner or f"{socket.gethostname()}:{os.getpid()}:{uuid.uuid4().hex}"
        key = self.key("lock", name)
        acquired = await self.redis.set(key, owner, nx=True, px=max(1, int(ttl_ms)))
        if not acquired:
            return None
        return RedisRuntimeLease(self.redis, key, owner, ttl_ms)

    async def close(self) -> None:
        close = getattr(self.redis, "aclose", None)
        if close is not None:
            await close()


def runtime_redis_url() -> str:
    return (os.getenv("RUNTIME_REDIS_URL") or os.getenv("ACCOUNT_REDIS_URL") or "").strip()


def create_runtime_store_from_env() -> RedisRuntimeStore | None:
    url = runtime_redis_url()
    if not url:
        return None
    from redis.asyncio import Redis

    return RedisRuntimeStore(Redis.from_url(url, decode_responses=False))


__all__ = [
    "RedisRuntimeLease",
    "RedisRuntimeStore",
    "create_runtime_store_from_env",
    "runtime_redis_url",
]

File: app/platform/runtime/redis_runtime.go
package runtime

import (
	"context"
	"fmt"
	"os"
	"strings"
	"time"

	redis "github.com/redis/go-redis/v9"
)

type RedisRuntimeClient interface {
	Get(ctx context.Context, key string) (any, error)
	SetNX(ctx context.Context, key, value string, ttlMS int) (bool, error)
	Expire(ctx context.Context, key string, ttlSeconds int) error
	Delete(ctx context.Context, key string) error
	CompareExpire(ctx context.Context, key string, owner string, ttlMS int) (bool, error)
	CompareDelete(ctx context.Context, key string, owner string) (bool, error)
}

type RedisRuntimeClientFactory func(rawURL string) (RedisRuntimeClient, error)

const defaultRedisRuntimeLockTTLMS = 300000

type RedisRuntimeStoreOptions struct {
	KeyPrefix string
}

type RedisRuntimeLockOptions struct {
	Owner string
	TTLMS int
}

type RedisRuntimeLease struct {
	client   RedisRuntimeClient
	Key      string
	Owner    string
	TTLMS    int
	released bool
}

type RedisRuntimeStore struct {
	Redis     RedisRuntimeClient
	KeyPrefix string
}

type goRedisRuntimeClient struct {
	client *redis.Client
}

func newGoRedisRuntimeClient(rawURL string) (RedisRuntimeClient, error) {
	options, err := redis.ParseURL(rawURL)
	if err != nil {
		return nil, err
	}
	return &goRedisRuntimeClient{client: redis.NewClient(options)}, nil
}

func NewRedisRuntimeLease(client RedisRuntimeClient, key, owner string, ttlMS int) *RedisRuntimeLease {
	return &RedisRuntimeLease{
		client: client,
		Key:    key,
		Owner:  owner,
		TTLMS:  maxRuntimeInt(1, ttlMS),
	}
}

func (l *RedisRuntimeLease) Renew(ctx context.Context) (bool, error) {
	if l.released {
		return false, nil
	}
	return l.client.CompareExpire(ctx, l.Key, l.Owner, l.TTLMS)
}

func (l *RedisRuntimeLease) Release(ctx context.Context) (bool, error) {
	if l.released {
		return false, nil
	}
	released, err := l.client.CompareDelete(ctx, l.Key, l.Owner)
	if err != nil {
		return false, err
	}
	l.released = true
	return released, nil
}

func NewRedisRuntimeStore(client RedisRuntimeClient, options RedisRuntimeStoreOptions) *RedisRuntimeStore {
	prefix := options.KeyPrefix
	if prefix == "" {
		prefix = "runtime:"
	}
	return &RedisRuntimeStore{Redis: client, KeyPrefix: prefix}
}

func (s *RedisRuntimeStore) Key(parts ...string) string {
	safe := []string{}
	for _, part := range parts {
		if part == "" {
			continue
		}
		safe = append(safe, strings.Trim(part, ":"))
	}
	return s.KeyPrefix + strings.Join(safe, ":")
}

func (s *RedisRuntimeStore) AcquireLock(ctx context.Context, name string, options RedisRuntimeLockOptions) (*RedisRuntimeLease, error) {
	owner := options.Owner
	if owner == "" {
		owner = runtimeOwner()
	}
	ttlMS := options.TTLMS
	if ttlMS == 0 {
		ttlMS = defaultRedisRuntimeLockTTLMS
	}
	ttlMS = maxRuntimeInt(1, ttlMS)
	key := s.Key("lock", name)
	acquired, err := s.Redis.SetNX(ctx, key, owner, ttlMS)
	if err != nil {
		return nil, err
	}
	if !acquired {
		return nil, nil
	}
	return NewRedisRuntimeLease(s.Redis, key, owner, ttlMS), nil
}

func (s *RedisRuntimeStore) Close(ctx context.Context) error {
	closer, ok := s.Redis.(interface {
		AClose(context.Context) error
	})
	if !ok {
		return nil
	}
	return closer.AClose(ctx)
}

func RuntimeRedisURL() string {
	if value := strings.TrimSpace(os.Getenv("RUNTIME_REDIS_URL")); value != "" {
		return value
	}
	return strings.TrimSpace(os.Getenv("ACCOUNT_REDIS_URL"))
}

func CreateRuntimeStoreFromEnv(factory RedisRuntimeClientFactory) (*RedisRuntimeStore, error) {
	rawURL := RuntimeRedisURL()
	if rawURL == "" {
		return nil, nil
	}
	if factory == nil {
		factory = newGoRedisRuntimeClient
	}
	client, err := factory(rawURL)
	if err != nil {
		return nil, err
	}
	return NewRedisRuntimeStore(client, RedisRuntimeStoreOptions{}), nil
}

func (c *goRedisRuntimeClient) Get(ctx context.Context, key string) (any, error) {
	value, err := c.client.Get(ctx, key).Bytes()
	if err == redis.Nil {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	return value, nil
}

func (c *goRedisRuntimeClient) SetNX(ctx context.Context, key, value string, ttlMS int) (bool, error) {
	ttl := time.Duration(maxRuntimeInt(1, ttlMS)) * time.Millisecond
	return c.client.SetNX(ctx, key, value, ttl).Result()
}

func (c *goRedisRuntimeClient) Expire(ctx context.Context, key string, ttlSeconds int) error {
	ttl := time.Duration(maxRuntimeInt(1, ttlSeconds)) * time.Second
	return c.client.Expire(ctx, key, ttl).Err()
}

func (c *goRedisRuntimeClient) Delete(ctx context.Context, key string) error {
	return c.client.Del(ctx, key).Err()
}

func (c *goRedisRuntimeClient) CompareExpire(ctx context.Context, key string, owner string, ttlMS int) (bool, error) {
	result, err := c.client.Eval(ctx, `
if redis.call("GET", KEYS[1]) == ARGV[1] then
  return redis.call("PEXPIRE", KEYS[1], ARGV[2])
end
return 0
`, []string{key}, owner, maxRuntimeInt(1, ttlMS)).Int()
	return result == 1, err
}

func (c *goRedisRuntimeClient) CompareDelete(ctx context.Context, key string, owner string) (bool, error) {
	result, err := c.client.Eval(ctx, `
if redis.call("GET", KEYS[1]) == ARGV[1] then
  return redis.call("DEL", KEYS[1])
end
return 0
`, []string{key}, owner).Int()
	return result == 1, err
}

func (c *goRedisRuntimeClient) HSet(ctx context.Context, key string, mapping map[string]string) error {
	return c.client.HSet(ctx, key, runtimeRedisHashArgs(mapping)...).Err()
}

func (c *goRedisRuntimeClient) HGetAll(ctx context.Context, key string) (map[string]any, error) {
	values, err := c.client.HGetAll(ctx, key).Result()
	if err != nil {
		return nil, err
	}
	result := make(map[string]any, len(values))
	for field, value := range values {
		result[field] = value
	}
	return result, nil
}

func (c *goRedisRuntimeClient) AClose(context.Context) error {
	return c.client.Close()
}

func runtimeRedisHashArgs(mapping map[string]string) []any {
	args := make([]any, 0, len(mapping)*2)
	for key, value := range mapping {
		args = append(args, key, value)
	}
	return args
}

func decodeRedisRuntimeValue(value any) (string, bool) {
	switch typed := value.(type) {
	case nil:
		return "", false
	case []byte:
		return string(typed), true
	case string:
		return typed, true
	default:
		return fmt.Sprint(typed), true
	}
}

func runtimeOwner() string {
	host, err := os.Hostname()
	if err != nil || host == "" {
		host = "localhost"
	}
	return fmt.Sprintf("%s:%d:%s", host, os.Getpid(), NextHex(32))
}

func maxRuntimeInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}
